### PR TITLE
Issue #4783: Prevent accidentally setting admin role

### LINKED ISSUE #4783: An arbitrary role can accidentally be set as admin when the role that was previously set as admin is deleted
I've noticed this while working on #3813

## Steps to reproduce:

1. Depending on what you have access to, either:
   1. Go to `admin/config/development/configuration/single/export` -> Select "Configuration" from the group dropdown -> then select "System core settings"
   1. Open your `system.core.json` config file
1. Inspect the `user_admin_role` setting and confirm that it is set to: `"user_admin_role": "administrator"` 👍 
1. Go to `admin/config/people/roles`
1. Expand the "Administrator role" fieldset -> confirm that the "Administrator" role is selected as admin 👍 
1. Add a new role
1. Expand the "Administrator role" fieldset -> set the newly-added role from the previous step as admin -> save
1. Expand the "Administrator role" fieldset -> confirm that the new role is now configured as admin 👍
1. Once again, either check your `system.core.json` config file, or in `admin/config/development/configuration/single/export` -> confirm that the `user_admin_role` setting is set to the machine name of the new role, instead of `"administrator"` 👍 
1. Back in `admin/config/people/roles` -> delete the role you have created (which is also set as admin).
1. Yet once again, either check your `system.core.json` config file, or in `admin/config/development/configuration/single/export` -> confirm that the `user_admin_role` setting is missing/deleted 👍 (perhaps that shouldn't happen and this is the actual culprit? ...but that's the way things work currently)
1. Back in `admin/config/people/roles` -> Expand the "Administrator role" fieldset -> ...

## Actual behavior

A random role is selected in the dropdown menu (whichever role is at the top of the list; usually "Editor") 👎 ...this means that if people delete a role that was previously set as admin, and then they reorder the roles and save the form (without checking the "Administrator role" fieldset, which is collapsed by default), they may set a random role as admin ...which means that this role will be granted ALL permissions from there on 👎 👎 👎 👎 

## Expected behavior

Things we could do:

1. Make sure that the "disabled" option is the first option in the dropdown menu, so that that is the one that gets selected by default.
1. When deleting the role that's set as admin:
   - throw a warning
   - set the "Administrator role" fieldset to be expanded instead of collapsed (although people may still miss this and accidentally set an arbitrary role as admin).
1. Make sure that after a role that was previously set as admin is deleted:
   - the admin dropdown is set to the "disabled" option (so that no random role gets accidentally set as admin)
   - instead of removing the `"user_admin_role"` entry in `system.core.json`, explicitly set it to `"user_admin_role": 0`
1. In the "Delete role" confirmation dialog:
   - show a warning if the to-be-deleted role is set as admin
   - provide a dropdown, to prompt selecting a new role to be set as admin ...or select the "disabled" option

### PR DESCRIPTION
**Pull request by klonos**
Fixes https://github.com/backdrop/backdrop-issues/issues/4783

### CODE REVIEW

**Comment by klonos** [comment]
...refreshing PR sandbox

**Comment by klonos** [comment]
...again 😓

File: core/modules/user/user.module [hunk 1]
```diff
@@ -2528,23 +2528,23 @@ function user_role_save($role) {
...
-  $config->delete();
+  $user_config = config('user.role.' . $role_name);
+  $user_config->delete();
 
   // Update the users who have this role set:
   db_delete('users_roles')
     ->condition('role', $role_name)
     ->execute();
 
+  $role = user_role_load($role_name);
```
Comment (indigoxela) [inline core/modules/user/user.module]: This confuses me.
Above you're using the "shortcut" to load the config directly instead of using user_role_load() - `config('user.role.' . $role_name);`

But here you're using user_role_load() again. On the role that has just been deleted. Mind to explain a bit?

File: core/modules/user/user.install [hunk 2]
```diff
@@ -1229,6 +1229,17 @@ function user_update_1018() {
   $config->save();
 }
 
+/**
+ * Explicitly set the "user_admin_role" attribute to "0" if it's missing.
```
Comment (indigoxela) [inline core/modules/user/user.install]: Maybe not important, but this is the text the admin user sees after running the update, right?
Maybe use "disabled" here instead of the ambiguous "0"? Just a suggestion.

File: core/modules/user/user.module (hunk 1 above)
Comment (klonos) [reply core/modules/user/user.module]: This is not actually me doing this - I'm just rearranging preexisting code. This line here was initially in line 2531, right after the function declaration.

File: core/modules/user/user.install (hunk 2 above)
Comment (klonos) [reply core/modules/user/user.install]: This is exactly what we're doing though 🤷 ...perhaps we could be more descriptive(?) ...perhaps something like this instead?:

```suggestion
 * Ensure that the admin role setting is disabled, by explicitly setting the
 * value of the "user_admin_role" attribute to "0" if it's missing.
```

File: core/modules/user/user.module (hunk 1 above)
Comment (indigoxela) [reply core/modules/user/user.module]: Right, but your rearrangement means that you load the role *after* it has been deleted. Before your change the role has been loaded *before* it got deleted.

This very likely breaks hook_user_role_delete (below).

File: core/modules/user/user.install (hunk 2 above)
Comment (indigoxela) [reply core/modules/user/user.install]: > This is exactly what we're doing though

That's correct. Your alternate suggestions seems a bit wordy to me. So maybe leave it as-is? Whatever you decide.

Comment (klonos) [reply core/modules/user/user.install]: Thanks @indigoxela 🙏 ...I'll wait for a few more people to review, and see what they think. I'll update this accordingly then.

File: core/modules/user/user.module (hunk 1 above)
Comment (klonos) [reply core/modules/user/user.module]: 🤔

Comment (klonos) [reply core/modules/user/user.module]: ...makes me think if `module_invoke_all()` below should be called earlier (before the role is deleted), or even if we should be passing the role name instead of the role object to the hook.

Anyway, it seems that this change should not be part of this PR. I'm usually good with adding inline comments when making changes that are required, and since I don't see any comments here, I don't think that this was required/intentional.

File: core/modules/user/user.install (hunk 2 above)
Comment (klonos) [reply core/modules/user/user.install]: I've reverted this change now.

File: core/modules/user/user.module (hunk 1 above)
Comment (klonos) [reply core/modules/user/user.module]: ...I've reverted that specific change now.

File: core/modules/user/user.install [hunk 3]
```diff
@@ -1229,6 +1229,18 @@ function user_update_1018() {
   $config->save();
 }
 
+/**
+ * Explicitly set the "user_admin_role" attribute to "0" (disabled) if it's
+ * missing.
```
Comment (quicksketch) [inline core/modules/user/user.install]: The first line of any function docblock should be a stand-alone sentence (followed by a blank line if there is more documentation). Let's shorten this up just by removing "Explicitly".

```suggestion
 * Set the "user_admin_role" attribute to "0" (disabled) if it's missing.
```

**Comment by quicksketch** [comment]
Merged into 1.x and 1.17.x. Thanks!
